### ISSUE TITLE
varnish 7.x command compatibility

### ISSUE THREAD

**Issue by gutmensch** [martin-helmich/go-varnish-client#27]
Dear Martin,
first of all: great project, thanks! I tried lately to use the module talking to varnish 7.3 and stumbled upon this problem
```
I0525 06:34:18.514868       1 roundtrip_impl.go:124] writing to server: vcl.state
I0525 06:34:18.514878       1 roundtrip_impl.go:19] writing 10 bytes to server: 76636c2e737461746520 ("vcl.state ")
I0525 06:34:18.514904       1 roundtrip_impl.go:19] writing 30 bytes to server: 2272656c6f61645f32303233303532355f3036333431375f343437303022 ("\"reload_20230525_063417_44700\"")
I0525 06:34:18.514921       1 roundtrip_impl.go:19] writing 1 bytes to server: 20 (" ")
I0525 06:34:18.514935       1 roundtrip_impl.go:19] writing 6 bytes to server: 22636f6c6422 ("\"cold\"")
I0525 06:34:18.514942       1 roundtrip_impl.go:19] writing 1 bytes to server: 0a ("\n")
I0525 06:34:18.514947       1 roundtrip_impl.go:156] request written; waiting for response
I0525 06:34:18.515005       1 roundtrip_impl.go:58] read 13 bytes of header
I0525 06:34:18.515014       1 roundtrip_impl.go:59] header: "100 94      \n"
I0525 06:34:18.515017       1 roundtrip_impl.go:71] received message from Varnish server: response code 100, body length 94
I0525 06:34:18.515020       1 roundtrip_impl.go:80] 95 bytes read
I0525 06:34:18.515026       1 roundtrip_impl.go:81] message body: "Unknown request.\nType 'help' for more info.\nSyntax Error: Missing separator between arguments\n\n"
I0525 06:34:18.515036       1 watch.go:195] error while changing state of VCL reload_20230525_063417_44700: error while setting VCL state (code 100): Unknown request.
Type 'help' for more info.
Syntax Error: Missing separator between arguments
```

This impacts at least the `SetVCLState` and `DefineInlineVCL` calls. IMO this is a varnish bug introduced with https://github.com/varnishcache/varnish-cache/commit/a2ab44decc08bee407b232d3ae97b9b281b56c60#diff-a1503af454177018bb731c118bd8dad34c34d94a93c2605e6a3ee3ab17bfe686R194 because it seems varnish *wrongly detects a double quote as separator now instead of the space* although go-varnish-client is actually escaping the arguments correctly AFAICS also from the debug log above.

Solutions:
1. I avoided inline vcl by writing a vcl file into /tmp and then just just LoadVCL
2. I removed the extra quoting for SetVCLState to avoid the wrong detection of double quote as separator https://github.com/martin-helmich/go-varnish-client/compare/master...gutmensch:go-varnish-client:master

None of both are really a good fix. But at least I wanted to document it here.
Cheers,
Robert

**Comment by martin-helmich** [comment]
Hey Robert! 👋   Thanks for the report, and your analysis. I agree that both solutions are suboptimal; it looks like dropping quoting entirely should certainly break stuff in some cases (at the very least, `DefineInlineVCL`).

It would be interesting to know what the original `varnishadm` does in these cases, since that almost certainly won't be broken with Varnish 7. Maybe it's time to read the original Varnish source code again (or to do some Wireshark'ing 😕).

**Comment by gutmensch** [comment]
@martin-helmich Ah thanks for the reminder, I actually tested this with varnishadm too (taking the generated values from kube-httpcache and applied them inline) and it worked as expected (like before with varnish 6.x), so yeah, I also wondered if related changes to varnishadm exist. Wireshark'ing sounds like a great idea to see what varnishadm does differently here. 🚀

**Comment by martin-helmich** [comment]
Well, good news: I've been able to reproduce this issue in a unit test (see https://github.com/martin-helmich/go-varnish-client/pull/29, or more precisely https://github.com/martin-helmich/go-varnish-client/actions/runs/5088338453/jobs/9144624188). Now just to fix it... 😅

**Comment by gutmensch** [comment]
Nice one :)

**Comment by martin-helmich** [comment]
I did some digging (using `strace`, Wireshark would have been overkill); when using quotes, `varnishadm` exhibits the same behaviour:

```
write(3, "vl.state \"boot\" \"auto\"", 22) = 22
write(3, "\n", 1)                       = 1
rt_sigprocmask(SIG_BLOCK, [HUP INT QUIT TERM CONT TSTP WINCH], [], 8) = 0
rt_sigaction(SIGINT, {sa_handler=0xffff9cb02420, sa_mask=[], sa_flags=0}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
rt_sigaction(SIGTSTP, {sa_handler=0xffff9cb02420, sa_mask=[], sa_flags=0}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
rt_sigaction(SIGQUIT, {sa_handler=0xffff9cb02420, sa_mask=[], sa_flags=0}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
rt_sigaction(SIGHUP, {sa_handler=0xffff9cb02420, sa_mask=[], sa_flags=0}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
rt_sigaction(SIGTERM, {sa_handler=0xffff9cb02420, sa_mask=[], sa_flags=0}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
rt_sigaction(SIGCONT, {sa_handler=0xffff9cb02420, sa_mask=[], sa_flags=0}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
rt_sigaction(SIGWINCH, {sa_handler=0xffff9cb02420, sa_mask=[], sa_flags=0}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
ioctl(0, TCGETS, {B38400 opost isig icanon echo ...}) = 0
ioctl(0, TCGETS, {B38400 opost isig icanon echo ...}) = 0
ioctl(0, SNDCTL_TMR_STOP or TCSETSW, {B38400 opost isig -icanon -echo ...}) = 0
ioctl(0, TCGETS, {B38400 opost isig -icanon -echo ...}) = 0
rt_sigprocmask(SIG_BLOCK, [WINCH], [], 8) = 0
ioctl(0, TIOCGWINSZ, {ws_row=59, ws_col=237, ws_xpixel=0, ws_ypixel=0}) = 0
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
write(1, "varnish> ", 9varnish> )                = 9
ppoll([{fd=3, events=POLLIN}, {fd=0, events=POLLIN}], 2, NULL, NULL, 0) = 1 ([{fd=3, revents=POLLIN}])
ppoll([{fd=3, events=POLLIN}], 1, {tv_sec=5, tv_nsec=0}, NULL, 0) = 1 ([{fd=3, revents=POLLIN}], left {tv_sec=4, tv_nsec=999819792})
read(3, "100 94      \n", 13)           = 13
ppoll([{fd=3, events=POLLIN}], 1, {tv_sec=5, tv_nsec=0}, NULL, 0) = 1 ([{fd=3, revents=POLLIN}], left {tv_sec=4, tv_nsec=999997500})
read(3, "Unknown request.\nType 'help' for"..., 95) = 95
100        r           \r100\n", 17
)    = 17
write(1, "Unknown request.\nType 'help' for"..., 94Unknown request.
Type 'help' for more info.
Syntax Error: Missing separator between arguments
) = 94
write(1, "\n", 1
)                       = 1
```

So for `vcl.state`, it might actually be prudent to just drop the quoting.

`vcl.inline` works in varnishadm when using quotes for the VCL itself; strace looks like this:

```
$ strace -s 4096 varnishadm vcl.inline test2 '"vcl 4.0; backend default { .host = \"127.0.0.1\"; }"' auto
write(3, "vcl.inline", 10)              = 10
write(3, " ", 1)                        = 1
write(3, "test2", 5)                    = 5
write(3, " ", 1)                        = 1
write(3, "\"vcl 4.0; backend default { .host = \\\"127.0.0.1\\\"; }\"", 53) = 53
write(3, " ", 1)                        = 1
write(3, "auto", 4)                     = 4
write(3, "\n", 1)                       = 1
ppoll([{fd=3, events=POLLIN}], 1, {tv_sec=5, tv_nsec=0}, NULL, 0) = 1 ([{fd=3, revents=POLLIN}], left {tv_sec=4, tv_nsec=820625042})
read(3, "200 0       \n", 13)           = 13
ppoll([{fd=3, events=POLLIN}], 1, {tv_sec=5, tv_nsec=0}, NULL, 0) = 1 ([{fd=3, revents=POLLIN}], left {tv_sec=4, tv_nsec=999999250})
read(3, "\n", 1)                        = 1
```

However, `vcl.inline` _also_ shows the same behaviour when the VCL state is supplied as a quoted string:

```
$ strace varnishadm vcl.inline test3 '"vcl 4.0; backend default { .host = \"127.0.0.1\"; }"' '"auto"'
...
write(3, "vcl.inline", 10)              = 10
write(3, " ", 1)                        = 1
write(3, "test3", 5)                    = 5
write(3, " ", 1)                        = 1
write(3, "\"vcl 4.0; backend default { .hos"..., 53) = 53
write(3, " ", 1)                        = 1
write(3, "\"auto\"", 6)                 = 6
write(3, "\n", 1)                       = 1
ppoll([{fd=3, events=POLLIN}], 1, {tv_sec=5, tv_nsec=0}, NULL, 0) = 1 ([{fd=3, revents=POLLIN}], left {tv_sec=4, tv_nsec=999999334})
read(3, "100 94      \n", 13)           = 13
ppoll([{fd=3, events=POLLIN}], 1, {tv_sec=5, tv_nsec=0}, NULL, 0) = 1 ([{fd=3, revents=POLLIN}], left {tv_sec=4, tv_nsec=999999292})
read(3, "Unknown request.\nType 'help' for"..., 95) = 95
fstat(1, {st_mode=S_IFCHR|0600, st_rdev=makedev(0x88, 0), ...}) = 0
write(1, "Unknown request.\n", 17Unknown request.
)      = 17
write(1, "Type 'help' for more info.\n", 27Type 'help' for more info.
) = 27
write(1, "Syntax Error: Missing separator "..., 50Syntax Error: Missing separator between arguments
) = 50
write(1, "\n", 1
)                       = 1
```

I think I'll just adjust the library to drop the quotes around the values that do not explicitly require them.

**Comment by gutmensch** [comment]
I just wondered isn't this exactly the case that _should_ actually work according to the test here https://github.com/varnishcache/varnish-cache/blob/master/lib/libvarnish/vav.c#L389 but it is still triggered through this condition https://github.com/varnishcache/varnish-cache/blob/master/lib/libvarnish/vav.c#L195 ? Or is it because they don't have a test with e.g. 1 unquoted arg and 2 quoted args?

**Comment by martin-helmich** [comment]
Honestly, mostly for reasons of time, I gave up on digging through the Varnish source code; I recently tagged a recent version of this library with my changes from #29 because I needed a v7-compatible version for the mittwald/kube-httpcache project (see mittwald/kube-httpcache#158); my first tests with the current HEAD of this library were promising.

I'll go ahead and close this issue as completed, for now. If you happen to stumble across any further insights as to why Varnish behaves the way it does in this case, feel free to add them here. 😉
